app/controllers/application_controller.rb

```diff
@@ -1,14 +1,18 @@
 # frozen_string_literal: true
 
 class ApplicationController < ActionController::Base
+  include ParamConverters
+
   before_action :authenticate_auth_user!
   around_action :switch_locale
 
   default_form_builder SkillsFormBuilder
 
   def switch_locale(&)
-    locale = params[:locale] || I18n.default_locale
-    I18n.with_locale(locale, &)
+    param_locale = params[:locale]
+    redirect_to(locale: cookies[:locale] || I18n.default_locale) unless param_locale
+    cookies.permanent[:locale] = param_locale if true?(params[:set_by_user])
```

Comment:
As `params[:set_by_user]` can only ever by true or nil you don't really need to do param conversion here.
Just check if the param is present or not.